From 88b2bb1c07e3c87fd326470d531bb0dfb2456f16 Mon Sep 17 00:00:00 2001
From: James Morse <james.morse@arm.com>
Date: Mon, 22 Jan 2024 14:58:12 +0800
Subject: [PATCH 660/960] downstream: arm_mpam: resctrl: Add CDP emulation

Intel RDT's CDP feature allows the cache to use a different control value
depending on whether the accesses was for instruction fetch or a data
access. MPAM's equivalent feature is the other way up: the CPU assigns a
different partid label to traffic depending on whether it was instruction
fetch or a data access, which causes the cache to use a different control
value based solely on the partid.

MPAM can emulate CDP, with the side effect that the alternative partid is
seen by all MSC, it can't be enabled per-MSC.

Add the resctrl hooks to turn this on or off. Add the helpers that
match a closid against a task, which need to be aware that the value
written to hardware is not the same as the one resctrl is using.

The context switch code needs to match the default resctrl group's
value against a variable, as this value changes depending on whether
CDP is in use.

Signed-off-by: James Morse <james.morse@arm.com>
Signed-off-by: Zeng Heng <zengheng4@huawei.com>
Signed-off-by: Liu Shixin <liushixin2@huawei.com>
---
 arch/arm64/include/asm/mpam.h        |  3 +-
 drivers/platform/mpam/mpam_resctrl.c | 62 ++++++++++++++++++++++++++++
 include/linux/arm_mpam.h             | 13 ++++++
 3 files changed, 77 insertions(+), 1 deletion(-)

diff --git a/arch/arm64/include/asm/mpam.h b/arch/arm64/include/asm/mpam.h
index 1d81a6f26acd..edae8c98fa28 100644
--- a/arch/arm64/include/asm/mpam.h
+++ b/arch/arm64/include/asm/mpam.h
@@ -4,6 +4,7 @@
 #ifndef __ASM__MPAM_H
 #define __ASM__MPAM_H
 
+#include <linux/arm_mpam.h>
 #include <linux/bitops.h>
 #include <linux/init.h>
 #include <linux/jump_label.h>
@@ -108,7 +109,7 @@ static inline void mpam_thread_switch(struct task_struct *tsk)
 	    !static_branch_likely(&mpam_enabled))
 		return;
 
-	if (!regval)
+	if (regval == READ_ONCE(mpam_resctrl_default_group))
 		regval = READ_ONCE(per_cpu(arm64_mpam_default, cpu));
 
 	oldregval = READ_ONCE(per_cpu(arm64_mpam_current, cpu));
diff --git a/drivers/platform/mpam/mpam_resctrl.c b/drivers/platform/mpam/mpam_resctrl.c
index 3d001af48c0c..075d8739efea 100644
--- a/drivers/platform/mpam/mpam_resctrl.c
+++ b/drivers/platform/mpam/mpam_resctrl.c
@@ -19,6 +19,8 @@
 
 #include "mpam_internal.h"
 
+u64 mpam_resctrl_default_group;
+
 /*
  * The classes we've picked to map to resctrl resources.
  * Class pointer may be NULL.
@@ -29,6 +31,12 @@ static bool exposed_alloc_capable;
 static bool exposed_mon_capable;
 static struct mpam_class *mbm_local_class;
 
+/*
+ * MPAM emulates CDP by setting different PARTID in the I/D fields of MPAM1_EL1.
+ * This applies globally to all traffic the CPU generates.
+ */
+static bool cdp_enabled;
+
 bool resctrl_arch_alloc_capable(void)
 {
 	return exposed_alloc_capable;
@@ -44,6 +52,36 @@ bool resctrl_arch_is_mbm_local_enabled(void)
 	return mbm_local_class;
 }
 
+bool resctrl_arch_get_cdp_enabled(enum resctrl_res_level ignored)
+{
+	return cdp_enabled;
+}
+
+int resctrl_arch_set_cdp_enabled(enum resctrl_res_level ignored, bool enable)
+{
+	u64 regval;
+	u32 partid, partid_i, partid_d;
+
+	cdp_enabled = enable;
+
+	partid = RESCTRL_RESERVED_CLOSID;
+
+	if (enable) {
+		partid_d = resctrl_get_config_index(partid, CDP_CODE);
+		partid_i = resctrl_get_config_index(partid, CDP_DATA);
+		regval = FIELD_PREP(MPAM_SYSREG_PARTID_D, partid_d) |
+			 FIELD_PREP(MPAM_SYSREG_PARTID_I, partid_i);
+
+	} else {
+		regval = FIELD_PREP(MPAM_SYSREG_PARTID_D, partid) |
+			 FIELD_PREP(MPAM_SYSREG_PARTID_I, partid);
+	}
+
+	WRITE_ONCE(mpam_resctrl_default_group, regval);
+
+	return 0;
+}
+
 /*
  * MSC may raise an error interrupt if it sees an out or range partid/pmg,
  * and go on to truncate the value. Regardless of what the hardware supports,
@@ -54,6 +92,30 @@ u32 resctrl_arch_get_num_closid(struct rdt_resource *ignored)
 	return min((u32)mpam_partid_max + 1, (u32)RESCTRL_MAX_CLOSID);
 }
 
+bool resctrl_arch_match_closid(struct task_struct *tsk, u32 closid)
+{
+	u64 regval = mpam_get_regval(tsk);
+	u32 tsk_closid = FIELD_GET(MPAM_SYSREG_PARTID_D, regval);
+
+	if (cdp_enabled)
+		tsk_closid >>= 1;
+
+	return tsk_closid == closid;
+}
+
+/* The task's pmg is not unique, the partid must be considered too */
+bool resctrl_arch_match_rmid(struct task_struct *tsk, u32 closid, u32 rmid)
+{
+	u64 regval = mpam_get_regval(tsk);
+	u32 tsk_closid = FIELD_GET(MPAM_SYSREG_PARTID_D, regval);
+	u32 tsk_rmid = FIELD_GET(MPAM_SYSREG_PMG_D, regval);
+
+	if (cdp_enabled)
+		tsk_closid >>= 1;
+
+	return (tsk_closid == closid) && (tsk_rmid == rmid);
+}
+
 struct rdt_resource *resctrl_arch_get_resource(enum resctrl_res_level l)
 {
 	if (l >= RDT_NUM_RESOURCES)
diff --git a/include/linux/arm_mpam.h b/include/linux/arm_mpam.h
index 0e4f94562ce6..163a26e1fc21 100644
--- a/include/linux/arm_mpam.h
+++ b/include/linux/arm_mpam.h
@@ -5,8 +5,17 @@
 #define __LINUX_ARM_MPAM_H
 
 #include <linux/acpi.h>
+#include <linux/resctrl_types.h>
 #include <linux/types.h>
 
+/*
+ * The value of the MPAM1_EL1 sysreg when a task is in the default group.
+ * This is used by the context switch code to use the resctrl CPU property
+ * instead. The value is modified when CDP is enabled/disabled by mounting
+ * the resctrl filesystem.
+ */
+extern u64 mpam_resctrl_default_group;
+
 struct mpam_msc;
 
 enum mpam_msc_iface {
@@ -53,4 +62,8 @@ static inline bool resctrl_arch_is_mbm_total_enabled(void)
 /* reset cached configurations, then all devices */
 void resctrl_arch_reset_resources(void);
 
+bool resctrl_arch_get_cdp_enabled(enum resctrl_res_level ignored);
+int resctrl_arch_set_cdp_enabled(enum resctrl_res_level ignored, bool enable);
+bool resctrl_arch_match_closid(struct task_struct *tsk, u32 closid);
+bool resctrl_arch_match_rmid(struct task_struct *tsk, u32 closid, u32 rmid);
 #endif /* __LINUX_ARM_MPAM_H */
-- 
2.43.0

